go/vt/sqlparser/ast.go

```diff
@@ -1766,6 +1768,51 @@ func (node *ExistsExpr) WalkSubtree(visit Visit) error {
 	)
 }
 
+// ExprFromValue converts the given Value into an Expr or returns an error.
+func ExprFromValue(value sqltypes.Value) (Expr, error) {
+	switch value.Type() {
+	case querypb.Type_NULL_TYPE:
+		return &NullVal{}, nil
+	case querypb.Type_INT8,
+		querypb.Type_UINT8,
+		querypb.Type_INT16,
+		querypb.Type_UINT16,
+		querypb.Type_INT24,
+		querypb.Type_UINT24,
+		querypb.Type_INT32,
+		querypb.Type_UINT32,
+		querypb.Type_INT64,
+		querypb.Type_UINT64:
+		return NewIntVal(value.ToBytes()), nil
+	case querypb.Type_FLOAT32,
+		querypb.Type_FLOAT64,
+		querypb.Type_DECIMAL:
+		return NewFloatVal(value.Raw()), nil
+	case querypb.Type_BIT:
+		return NewBitVal(value.Raw()), nil
+	case querypb.Type_TIMESTAMP,
+		querypb.Type_DATE,
+		querypb.Type_TIME,
+		querypb.Type_DATETIME,
+		querypb.Type_YEAR,
+		querypb.Type_TEXT,
+		querypb.Type_BLOB,
+		querypb.Type_VARCHAR,
+		querypb.Type_VARBINARY,
+		querypb.Type_CHAR,
+		querypb.Type_BINARY,
+		querypb.Type_ENUM,
+		querypb.Type_SET,
+		querypb.Type_TUPLE,
+		querypb.Type_GEOMETRY,
+		querypb.Type_JSON,
+		querypb.Type_EXPRESSION:
+		return NewStrVal(value.Raw()), nil
+	default:
+		return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "default value %v of type %v does not map to a sqlparser type", value.String(), value.Type())
+	}
```

Comment:
I actually think that the Bit to BitVal conversion may be incorrect. So, I'll need to look at how a Bit Value gets built and how it gets converted back. But I don't have time right now. I'll verify this later.

Reply:
Based on this:
```
mysql> create table a(id int, val bit(8) default b'101', primary key(id));
Query OK, 0 rows affected (0.02 sec)

mysql>
mysql> describe a;
+-------+---------+------+-----+---------+-------+
| Field | Type    | Null | Key | Default | Extra |
+-------+---------+------+-----+---------+-------+
| id    | int(11) | NO   | PRI | 0       |       |
| val   | bit(8)  | YES  |     | b'101'  |       |
+-------+---------+------+-----+---------+-------+
2 rows in set (0.00 sec)
```
We have two problems:
1. The sqltypes.Value we build for bit fields is incorrect here: https://github.com/youtube/vitess/blob/1595283c1e48b4c80ead69245292722e0ad4efc0/go/vt/vttablet/tabletserver/schema/schema.go#L159.
2. The SQLVal we build here should just be a string representation, likely escaped because the data will be binary. Basically, the value should be treated like a regular quoted string.

There is also a more concise and efficient way to do this without the elaborate switch.

I'll patch this PR and add a new commit to make the above fixes.

Reply:
I ran an audit for default values for all data types. Only the `bit` type has a representation that is different from what mysql would otherwise return as a value. The more exotic `geometry` and `json` types don't allow default values. So, I'm thinking of just special-casing this.

Also, we don't have code to convert something like `b'101'` to the binary representation. So, I'm thinking of making mysql do this for us with `select b'101'`.